Builtin Date_Time, Time_Of_Day, Zone

hubertp opened the pull request:
### Pull Request Description

Improved polyglot support for Date_Time (formerly Time), Time_Of_Day and
Zone. This follows the pattern introduced for Enso Date.

Related to https://www.pivotaltracker.com/story/show/182973664 and https://www.pivotaltracker.com/n/projects/2539304/stories/182987210

### Important Notes

Minor caveat - in tests for Date, had to bend a lot for JS Date to pass.
This is because JS Date is not really only a Date, but also a Time and
Timezone, previously we just didn't consider the latter.
Also, JS Date does not deal well with setting timezones so the trick I
used is to first call foreign function returning a polyglot JS Date,
which is converted to ZonedDateTime and only then set the correct
timezone. That way none of the existing tests had to be changed or
special cased.

Additionally, JS deals with milliseconds rather than nanoseconds so
there is loss in precision, as noted in Time_Spec.

### Checklist

Please include the following checklist in your PR:

- [ ] The documentation has been updated if necessary.
- [ ] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [ ] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

hubertp committed:
Builtin Date_Time, Time_Of_Day, Zone

Improved polyglot support for Date_Time (formerly Time), Time_Of_Day and
Zone. This follows the pattern introduced for Enso Date.

Minor caveat - in tests for Date, had to bend a lot for JS Date to pass.
This is because JS Date is not really only a Date, but also a Time and
Timezone, previously we just didn't consider the latter.
Also, JS Date does not deal well with setting timezones so the trick I
used is to first call foreign function returning a polyglot JS Date,
which is converted to ZonedDateTime and only then set the correct
timezone. That way none of the existing tests had to be changes or
special cased.

Additionally, JS deals with milliseconds rather than nanoseconds so
there is loss in precision, as noted in Time_Spec.

hubertp committed:
Add tests for Java's LocalTime

hubertp committed:
changelog

hubertp committed:
Make date formatters in table happy

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoZone.java:
```diff
@@ -0,0 +1,119 @@
+@ExportLibrary(MethodDispatchLibrary.class)
+@Builtin(pkg = "date", name = "Zone", stdlibName = "Standard.Base.Data.Time.Zone")
+public class EnsoZone implements TruffleObject {
+  private ZoneId zone;
```
I'd make the field `final`, if possible.

JaroslavTulach commented on test/Table_Tests/src/Data_Formatter_Spec.enso:
```diff
@@ -140,14 +140,14 @@ spec =
-            formatter.format (Time.new 1999 . internal_zoned_date_time) . should_equal "1999-01-01 00:00:00"
-            formatter.format (Time_Of_Day.new . internal_local_time) . should_equal "00:00:00"
+            formatter.format (Time.new 1999) . should_equal "1999-01-01 00:00:00"
+            formatter.format (Time_Of_Day.new) . should_equal "00:00:00"
```
Removal of `internal_xyz` is the desired achievement.

JaroslavTulach commented on std-bits/table/src/main/java/org/enso/table/formatting/TimeFormatter.java:
```diff
@@ -26,6 +33,6 @@ public String format(Object value) {
   @Override
   public boolean canFormat(Object value) {
-    return value instanceof LocalTime;
+    return value instanceof LocalTime || (value instanceof Value v && !v.isDate() && v.isTime());
```
Looks familiar. I played with something similar.

hubertp replied:
Yeah, I don't see other way around it.

JaroslavTulach commented on std-bits/table/src/main/java/org/enso/table/data/table/Column.java:
```diff
@@ -120,11 +120,23 @@ public Column rename(String name) {
     return new Column(name, new DefaultIndex(items.size()), builder.seal());
   }
 
+  private static Object convertDateOrTime(Value item) {
```
Correct. Values that are both date and time need this kind of care.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/type/Types.java:
```diff
@@ -221,6 +224,10 @@ private static TypeGraph buildTypeHierarchy() {
     graph.insert(ConstantsGen.PANIC, ConstantsGen.ANY);
     graph.insert(ConstantsGen.REF, ConstantsGen.ANY);
     graph.insert(ConstantsGen.TEXT, ConstantsGen.ANY);
+    graph.insert(ConstantsGen.DATE, ConstantsGen.ANY);
```
Looks like an omission from my earlier integration, right? What gets broken when the `graph` isn't up-to-date?

hubertp replied:
Nothing, apparently :shrug:

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/resolver/HostMethodCallNode.java:
```diff
@@ -59,7 +79,13 @@ public enum PolyglotCallType {
      */
     public boolean isInteropLibrary() {
-      return this != NOT_SUPPORTED && this != CONVERT_TO_TEXT && this != CONVERT_TO_DATE;
+      return this != NOT_SUPPORTED
```
The list is getting longer. A time will may have arrived to give the enum `final boolean interopLibrary` field and constructor.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/resolver/HostMethodCallNode.java:
```diff
@@ -80,8 +106,15 @@ public boolean isInteropLibrary() {
-    if (library.isDate(self) && !library.isTime(self)) {
-      return PolyglotCallType.CONVERT_TO_DATE;
+    if (library.isDate(self)) {
+      if (library.isTime(self))
```
Nitpick: I'd use `{` braces.

hubertp replied:
Sure. Other nitpick could be that this pattern of `isDate`, `isTime`, `isTimezone` is repeated all over the place. Not a fan of this solution but at least there aren't more cases.

JaroslavTulach commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Zone.enso:
```diff
@@ -120,7 +121,10 @@ type Zone
-    zone_id self = self.internal_zone_id . getId
+    zone_id self = @Builtin_Method "Zone.zone_id"
+
+    equals : Zone -> Boolean
```
Why there is `equal` as well as `==`? Something related to compatibility with former `asGuestObject(...).equals(...)`? Shall not such compatibility layer be ultimately removed?

hubertp replied:
You are right, `equals` is not necessary anymore

JaroslavTulach reviewed (approved):
Looks great to me overall. Some observations inline.

hubertp committed:
PR review, add more tests for zone

hubertp committed:
More tests and fixed a bug in column reader

Column reader didn't take into account timezone but that was a mistake
since then it wouldn't map to Enso's Date_Time.
Added tests that check it now.

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time.enso:
```diff
@@ -20,10 +20,10 @@ polyglot java import org.enso.base.Time_Utils
          example_now = Time.now
-now : Time
-now = Time ZonedDateTime.now
+now : DateTime
```
```suggestion
now : Date_Time
```

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time.enso:
```diff
@@ -358,7 +373,23 @@ type Time
+
+             example_to_time = Time.new 2020 2 3 . to_time Time_Of_Day.new Zone.utc
+    to_time : Time_Of_Day -> Zone -> Date_Time
+    to_time self time_of_day (zone=Zone.system) = self.to_time_builtin time_of_day zone
```
Something feels weird here. This function is defined on `Date_Time` and it returns a `Date_Time`. What is going on? This function would make sense on just `Date`. Alternatively, I guess we can have a method which replaces the time of day with a new one, keeping the same date, but I don't think calling such operation `to_time` will be understandable.

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Time_Of_Day.enso:
```diff
@@ -114,12 +114,9 @@ new (hour = 0) (minute = 0) (second = 0) (nanosecond = 0) =
-            formatter = DateTimeFormatter.ofPattern pattern
-            LocalTime.parse text (formatter.withLocale locale.java_locale)
-        _ -> Panic.throw (Time.Time_Error "An invalid pattern was provided.")
-    Time_Of_Day result . map_error <| case _ of
+        Nothing -> Time_Of_Day.parse_builtin text
+        Text    -> Time_Utils.parse_time text pattern locale.java_locale
+    result . map_error <| case _ of
         Polyglot_Error err -> Time.Time_Error err.getMessage
         x -> x
```
I think ideally we'd replace the `Panic.recover`+`map_error` with `Panic.catch_java`, as I think this is our go-to pattern for handling Java exceptions.

hubertp replied:
I will give it a try

radeusgd commented on distribution/lib/Standard/Table/0.0.0-dev/src/Data/Data_Formatter.enso:
```diff
@@ -171,14 +171,13 @@ type Data_Formatter
-                if datatype == Time then self.make_datetime_parser else
-                    if datatype == Time_Of_Day then self.make_time_parser else
-                        Error.throw (Illegal_Argument_Error "Unsupported datatype: "+datatype.to_text)
+        Integer     -> self.make_integer_parser
+        Decimal     -> self.make_decimal_parser
+        Boolean     -> self.make_boolean_parser
+        Date        -> self.make_date_parser
+        Date_Time   -> self.make_datetime_parser
+        Time_Of_Day -> self.make_time_parser
+        _           -> Error.throw (Illegal_Argument_Error "Unsupported datatype: "+datatype.to_text)
 
```
:heart:

hubertp committed:
remove redundant conversion

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java:
```diff
@@ -0,0 +1,274 @@
+    return dateTime.getYear();
+  }
+
+  @Builtin.Method(description = "Gets a value month")
```
```suggestion
  @Builtin.Method(description = "Gets the month")
```
Maybe something like this?

radeusgd replied:
(And same for others)

hubertp committed:
Update distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time.enso

Co-authored-by: Radosław Waśko <radoslaw.wasko@enso.org>

radeusgd commented on std-bits/base/src/main/java/org/enso/base/Time_Utils.java:
```diff
@@ -68,14 +62,46 @@ public static LocalDate date_adjust(LocalDate date, long add, Period duration) {
   }
 
-  public static long week_of_year(LocalDate date, Locale locale) {
+  public static ZonedDateTime datetime_adjust(
+      ZonedDateTime datetime, long add, Period period, Duration duration) {
+    return add == 1 ? datetime.plus(period).plus(duration) : datetime.minus(period).minus(duration);
+  }
```
I'm a bit hesitant of the usage of `long add` like this. If I set `add` to `0` or `100` it will have the same effect as `-1` - not very obvious.

Maybe use a `boolean forward` flag instead? Or ideally maybe we can just create an `enum Direction` for this?

hubertp replied:
I followed the pattern used in Date by @JaroslavTulach but sure, I can tweak it a bit

radeusgd commented on test/Table_Tests/src/Table_Time_Of_Day_Spec.enso:
```diff
@@ -0,0 +1,53 @@
+            table.columns.map (.name) . should_equal ['Serial number','Movement type', 'Posting time']
+            table.row_count.should_equal 6
+
+        Test.specify "should be able to treat a single value as a Date" <|
```
```suggestion
        Test.specify "should be able to treat a single value as a Time_Of_Day" <|
```

radeusgd commented on test/Table_Tests/src/Table_Time_Spec.enso:
```diff
@@ -0,0 +1,52 @@
+            table.columns.map (.name) . should_equal ['Serial number','Movement type', 'Posting date']
+            table.row_count.should_equal 6
+
+        Test.specify "should be able to treat a single value as a Date" <|
```
```suggestion
        Test.specify "should be able to treat a single value as a Time" <|
```

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time.enso:
```diff
@@ -121,47 +121,44 @@ new year (month = 1) (day = 1) (hour = 0) (minute = 0) (second = 0) (nanosecond
         x -> x
 
-type Time
+type Date_Time
```
Hm, shouldn't this be moved to a module called `Date_Time`?

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time.enso:
```diff
@@ -53,7 +53,7 @@ now = Time ZonedDateTime.now
          example_new = Time.new 1986 8 5
 new : Integer -> Integer -> Integer -> Integer -> Integer -> Integer -> Integer -> Zone -> Time ! Time_Error
```
```suggestion
new : Integer -> Integer -> Integer -> Integer -> Integer -> Integer -> Integer -> Zone -> Date_Time ! Time_Error
```

I think we should rename this module to `Date_Time`.

Currently when using this `new` function one would write `Time.new` but get type called `Date_Time` - IMO its a bit confusing, no?

radeusgd commented on std-bits/table/src/main/java/org/enso/table/formatting/TimeFormatter.java:
```diff
@@ -17,6 +19,11 @@ public String format(Object value) {
+    if (value instanceof Value v && v.isTime()) {
+      value = v.asTime();
+    }
+
+
```
```suggestion

```

radeusgd commented on test/Table_Tests/src/Table_Time_Spec.enso:
```diff
@@ -0,0 +1,52 @@
+
+from project.Util import all
+
+spec =
```
Should we have one more suite like that for `Date`?

hubertp replied:
There is one already

radeusgd replied:
Ah sorry, I just looked at the changes and since it was unchanged it didn't show up

radeusgd commented on test/Table_Tests/src/Data_Formatter_Spec.enso:
```diff
@@ -137,14 +137,14 @@ spec =
             formatter.format (Date.new 2022) . should_equal "2022-01-01"
-            formatter.format (Time.new 1999 . internal_zoned_date_time) . should_equal "1999-01-01 00:00:00"
-            formatter.format (Time_Of_Day.new . internal_local_time) . should_equal "00:00:00"
+            formatter.format (Time.new 1999) . should_equal "1999-01-01 00:00:00"
```
I'd appreciate also a test where we overwrite the zone to a different one, preferably two different ones (let's say CEST and PDT or something like these), to verify that the formatter correctly deals with zoned dates.

hubertp committed:
First round of addressing PR review

hubertp committed:
don't leak java exceptions in Zone

hubertp committed:
Move Date_Time to top-level module

hubertp commented:
@radeusgd I think I addressed all your comments

JaroslavTulach committed:
Merge branch 'develop' into wip/hubert/datetime-182973664

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/resolver/HostMethodCallNode.java:
```diff
@@ -48,6 +48,26 @@ public enum PolyglotCallType {
+     * The method call should be handled by converting {@code self} to a {@code
+     * Standard.Base.Data.DateTime} and dispatching natively.
+     */
+    CONVERT_TO_ZONEDDATETIME,
```
Nitpick: it should probably be `CONVERT_TO_ZONED_DATE_TIME` when the class is called `ZonedDateTime`. Similar for other constants. Not that it would affect the functionality, however.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java:
```diff
@@ -0,0 +1,274 @@
+@ExportLibrary(InteropLibrary.class)
+@ExportLibrary(MethodDispatchLibrary.class)
+@Builtin(pkg = "date", name = "DateTime", stdlibName = "Standard.Base.Data.Time.Date_Time")
+public class EnsoDateTime implements TruffleObject {
```
Could be `final`. Not that it would affect the runtime speed, however.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java:
```diff
@@ -0,0 +1,274 @@
+  }
+
+  @Builtin.Method(description = "Return the number of seconds from the Unix epoch.")
+  public long toEpochSeconds() {
```
My [recent experience](https://github.com/enso-org/enso/pull/3638/files#diff-07942731a467d64d7df16571e777af374f8b72159502eae8a436ba7addca768d) with `@TruffleBoundary` suggests that many of these methods shall be behind the boundary.

hubertp replied:
I can add some, but at the moment without any gatekeepers it feels like a slightly random process.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java:
```diff
@@ -0,0 +1,274 @@
+    }
+  }
+
+  @ExportMessage
```
This method must be behind `@TruffleBoundary` for sure.

hubertp committed:
PR review

jdunkerley reviewed (approved):
Thanks for this @hubertp - this gives us a fantastic base to build on to build out the functions we need.
